zio-http/src/main/scala/zio/http/Headers.scala

```diff
@@ -1,81 +1,151 @@
-    val (exceptions, regularHeaders) = self.toList.span(h => h._1.contains(HeaderNames.setCookie))
-    val combinedHeaders              = regularHeaders
-      .groupBy(_._1)
-      .map { case (key, tuples) =>
-        key -> tuples.map(_._2).mkString(",")
-      }
-    (exceptions ++ combinedHeaders)
-      .foldLeft[HttpHeaders](new DefaultHttpHeaders(true)) { case (headers, entry) =>
-        headers.add(entry._1, entry._2)
-      }
-  }
+  final def when(cond: Boolean): Headers = if (cond) self else Headers.EmptyHeaders
 
 }
 
 object Headers extends HeaderConstructors {
 
-  val empty: Headers   = Headers(Nil)
+  final case class Header private[zio] (key: CharSequence, value: CharSequence)
+      extends Product2[CharSequence, CharSequence]
+      with Headers {
+    self =>
+
+    override def _1: CharSequence = key
+
+    override def _2: CharSequence = value
+
+    override private[http] def encode: HttpHeaders = Headers.encode(self.toList)
+
+    def flatten: Iterable[Header] = Iterable(self)
+
+    override def iterator: Iterator[Header] =
+      Iterator.single(self)
+
+    // Unless we implement it headers.toString results in StackOverflow
+    override def toString(): String = (key, value).toString()
+
+  }
+
+  final case class FromIterable private[zio] (iter: Iterable[Header]) extends Headers with Iterable[Header] {
+    self =>
+
+    private[http] def encode: HttpHeaders = Headers.encode(self.toList)
+
+    def flatten: Iterable[Header] = iter.flatMap(_.flatten)
+
+    override def iterator: Iterator[Header] =
+      iter.iterator.flatMap(_.iterator)
+
+  }
+
+  final case class FromJHeaders private[zio] (toJHeaders: HttpHeaders) extends Headers {
+    self =>
+
+    override private[http] def encode: HttpHeaders = toJHeaders
+
+    def flatten: Iterable[Header] =
+      toJHeaders.entries().asScala.map(e => Header(e.getKey, e.getValue))
+
+    override def iterator: Iterator[Header] =
+      toJHeaders.entries().asScala.map(e => Header(e.getKey, e.getValue)).iterator
+
+  }
+
+  final case class Concat private[zio] (first: Headers, second: Headers) extends Headers {
+    self =>
+
+    override private[http] def encode: HttpHeaders = Headers.encode(self.toList)
+
+    override def flatten: Iterable[Header] = first.flatten ++ second.flatten
+
+    override def iterator: Iterator[Header] =
+      first.iterator ++ second.iterator
+
+  }
+
+  case object EmptyHeaders extends Headers {
+    self =>
+
+    override private[http] def encode: HttpHeaders = new DefaultHttpHeaders()
+
+    override def flatten: Iterable[Header] = Iterable.empty
+
+    override def iterator: Iterator[Header] =
+      Iterator.empty
+
+  }
+
   val BasicSchemeName  = "Basic"
   val BearerSchemeName = "Bearer"
 
-  def apply(name: CharSequence, value: CharSequence): Headers = Headers(Chunk((name, value)))
+  def apply(name: CharSequence, value: CharSequence): Headers = Headers.Header(name.toString, value.toString)
```

Comment:
The CorsSpec tests fail without the toString because the headers from HeadersConstructors have quoted values and retain them when passed as keys without the toString.  The alternative was to apply a toString on the headers in HeaderConstructors.